airbyte-server/src/main/java/io/airbyte/server/helpers/WorkspaceHelper.java

```diff
@@ -0,0 +1,155 @@
+      }
+
+    });
+
+    this.destinationToWorkspaceCache = getExpiringCache(new CacheLoader<>() {
+
+      @Override
+      public UUID load(UUID destinationId) throws JsonValidationException, ConfigNotFoundException, IOException {
+        final DestinationRead destination = destinationHandler.getDestination(new DestinationIdRequestBody().destinationId(destinationId));
+        return destination.getWorkspaceId();
+      }
+
+    });
+
+    this.connectionToWorkspaceCache = getExpiringCache(new CacheLoader<>() {
+
+      @Override
+      public UUID load(UUID connectionId) throws JsonValidationException, ConfigNotFoundException, IOException, ExecutionException {
+        final ConnectionRead connection = connectionsHandler.getConnection(new ConnectionIdRequestBody().connectionId(connectionId));
+        final UUID sourceId = connection.getSourceId();
+        final UUID destinationId = connection.getDestinationId();
+        return getWorkspaceForConnection(sourceId, destinationId);
+      }
+
+    });
+
+    this.jobToWorkspaceCache = getExpiringCache(new CacheLoader<>() {
+
+      @Override
+      public UUID load(Long jobId) throws IOException, ExecutionException {
+        final Job job = jobPersistence.getJob(jobId);
+        if (job.getConfigType() == JobConfig.ConfigType.SYNC || job.getConfigType() == JobConfig.ConfigType.RESET_CONNECTION) {
+          return getWorkspaceForConnectionId(UUID.fromString(job.getScope()));
+        } else {
+          throw new IllegalArgumentException("Only sync/reset jobs are associated with workspaces! A " + job.getConfigType() + " job was requested!");
+        }
+      }
+
+    });
+  }
+
+  public UUID getWorkspaceForSourceId(UUID sourceId) throws ExecutionException {
+    return sourceToWorkspaceCache.get(sourceId);
+  }
+
+  public UUID getWorkspaceForDestinationId(UUID destinationId) throws ExecutionException {
+    return destinationToWorkspaceCache.get(destinationId);
+  }
+
+  public UUID getWorkspaceForJobId(Long jobId) throws IOException, ExecutionException {
+    return jobToWorkspaceCache.get(jobId);
+  }
+
+  public UUID getWorkspaceForConnection(UUID sourceId, UUID destinationId) throws ExecutionException {
+    final UUID sourceWorkspace = getWorkspaceForSourceId(sourceId);
+    final UUID destinationWorkspace = getWorkspaceForDestinationId(destinationId);
+
+    Preconditions.checkArgument(Objects.equals(sourceWorkspace, destinationWorkspace), "Source and destination must be from the same workspace!");
+    return sourceWorkspace;
+  }
+
+  public UUID getWorkspaceForConnectionId(UUID connectionId) throws ExecutionException {
+    return connectionToWorkspaceCache.get(connectionId);
+  }
+
+  public UUID getWorkspaceForOperationId(UUID operationId) {
+    throw new NotImplementedException();
+  }
+
+  private static <K, V> LoadingCache<K, V> getExpiringCache(CacheLoader<K, V> cacheLoader) {
+    return CacheBuilder.newBuilder()
+        .expireAfterAccess(5, TimeUnit.MINUTES)
```

Comment:
what's the merit of a temporal expiration? i think all of these cases are where these ids should never change. if there is some reason that they are changing it's going to be kinda big deal and a 5 minute expiration is not really going to help. 

my sense is we probably don't want a temporal expiration? after that we can decide if we need manual expiration or not expiration. 

what do you think? am i way off here?

Reply:
I was mostly concerned about an infinitely growing cache where we don't want to keep all of the connections in memory. I switched it in https://github.com/airbytehq/airbyte/pull/4868/commits/d080a236f2da7277e148fdff2e1cece3d49d6d8e to a fixed 20k limit.

Does that sound good?

Reply:
i dig it. thanks!